### PR TITLE
Allow unused underscore identifiers

### LINKED ISSUE #361: Names prefixed with underscore shouldn't get 'unused name' diagnostic
That's at least what `rustc` does; I haven't looked up any details.

For example, as present `rustc` doesn't print any diagnostics, but with the underscore removed from `_fib` in `gcc/testsuite/rust.test/compile/loop5.rs`, it prints:

```
warning: unused variable: `fib`
 --> gcc/testsuite/rust.test/compile/loop5.rs:6:9
  |
6 |     let fib = loop {
  |         ^^^ help: if this is intentional, prefix it with an underscore: `_fib`
  |
  = note: `#[warn(unused_variables)]` on by default
```

We currently print the `unused name` diagnostic for both variants.

### PR DESCRIPTION
**Pull request by CohenArthur**
This fixes #361

### CODE REVIEW

**Comment by philberty** [comment]
This looks good to me. Though to me this looks like either 1 commit or two commits. I think two commits makes sense:

1. Add NodeId to CanonicalPath mappings (this is actually useful for name mangling later on)
2. Check for underscore before warning.

**Comment by CohenArthur** [comment]
> 
> 
> This looks good to me. Though to me this looks like either 1 commit or two commits. I think two commits makes sense:
> 
>     1. Add NodeId to CanonicalPath mappings (this is actually useful for name mangling later on)
> 
>     2. Check for underscore before warning.

Gotcha, I'll rebase right away

File: gcc/rust/resolve/rust-ast-resolve-unused.h [hunk 1]
```diff
@@ -30,7 +30,14 @@ class ScanUnused
   static void ScanRib (Rib *r)
   {
     r->iterate_decls ([&] (NodeId decl_node_id, Location locus) -> bool {
-      if (!r->have_references_for_node (decl_node_id))
+      CanonicalPath ident = CanonicalPath::create_empty ();
+      if (!r->lookup_id (decl_node_id, &ident))
+	{
+	  return false;
```
Comment (philberty) [inline gcc/rust/resolve/rust-ast-resolve-unused.h:36]: Can you put an assertion that this returns ok? Like:

```
bool ok = r->lookup_id();
rust_assert(ok);
```

 If this breaks the make check-rust then leave it as is.

**Review by philberty — APPROVED** [review]

**Review by philberty — APPROVED** [review]

**Pull request merged**